Some initial rebranch fixes

eeckstein opened the pull request:
For details see commit messages.

Status: Compiler + stdlib build successfully

eeckstein committed:
cmake: llvm now requires C++17

eeckstein committed:
fix wrong attribute ordering in the runtime

This fix might be replaced by something better, e.g. https://github.com/apple/swift/pull/61476

eeckstein committed:
replace `LLVM_NODISCARD` -> `[[nodiscard]]`

This is possible because we are now compiling with the C++17 standard.

eeckstein committed:
IRGen: remove the legacy pass manager

eeckstein committed:
AST: fix DeclContext's `cast_convert_val`

Replace it with `CastInfo`.
This fixes a build error with the new LLVM.

eeckstein committed:
LLVMPasses: fix SwiftAAResult

TODO:  how to construct a SimpleAAQueryInfo without an AAResults?

eeckstein committed:
Syntax: `llvm::in_place_t` -> `std::in_place_t`

eeckstein committed:
`llvm::array_lengthof` -> `std::size`

eeckstein committed:
adapt changed API for `clang::createInvocation`

eeckstein committed:
IRgen: `AllocAInst::getAlignment` -> `AllocaInst::getAlign`

eeckstein committed:
work around removal of `APInt::zextOrSelf` and `APInt::sextOrSelf`

eeckstein committed:
IRGen: remove a not needed an not available include

eeckstein committed:
IRGen: rename some LLVM passes

eeckstein committed:
IRGen: workaround a build error due to changed LTO passes

TODO: fix this

IRGen: Adjust lto/thin-lto to the new pass manager way of doing things

eeckstein committed:
JIT: `getAddress` -> `getValue`

eeckstein committed:
stdlib: adapt to llvm's changed `Optional` API

add `value()` which replaces the deprecated `getValue()`

edymtt committed:
[rebranch] Adapt to new signature for `llvm_install_symlink`...

and match the usage pattern employed by other LLVM projects.

For context about the underlying change see  https://reviews.llvm.org/D117977

Addresses rdar://101396797

(cherry picked from #61769, commit 2763cc3911705ab76c6966bc447f2667248b0ac3)

edymtt commented:
With regard to f6a5c08ec63d0ec3205f29982b6fcd4dd4df771a, we have a more fleshed out change in #61818 that should ensure the symlinks are generated correctly during installation.

bnbarham commented on lib/ClangImporter/ClangImporter.cpp:
```diff
@@ -1041,8 +1041,12 @@ std::unique_ptr<clang::CompilerInvocation> ClangImporter::createClangInvocation(
+    CIOpts.Diags = clangDiags;
+    CIOpts.RecoverOnError = false;
+    CIOpts.CC1Args = CC1Args;
+    CI = createInvocation(invocationArgs, std::move(CIOpts));
```
Should this be `clang::createInvocation`?

eeckstein committed:
ClangImporter: check if a decl's destructor is null

For some reason even if `decl->hasUserDeclaredDestructor()` is true, `decl->getDestructor()` still can be null.

eeckstein committed:
Workaround to avoid a compiler error because `cas::ObjectRef` is not defined when including VirtualFileSystem.h

bnbarham commented on include/swift/Basic/BasicBridging.h:
```diff
@@ -13,6 +13,10 @@
 #ifndef SWIFT_BASIC_BASICBRIDGING_H
 #define SWIFT_BASIC_BASICBRIDGING_H
 
+// Workaround to avoid a compiler error because `cas::ObjectRef` is not defined
+// when including VirtualFileSystem.h
+#include "llvm/CAS/CASReference.h"
+
```
What's requiring the actual definition? All the references I see are in `Optional`, which I thought would be fine with just a forward declaration. Is that not the case?

akyrtzi replied:
This is due to some issue with Swift/C++ interop, see rdar://98579497

akyrtzi replied:
We should cherry-pick `f93fb9e6953e12cd8f5eda6be9cd3dcd61de261f` in `stable/20221013`, but I'm still hesitant to put it in `next`, I think this is a bug with C++ interop that needs to be fixed.

bnbarham replied:
https://github.com/apple/llvm-project/pull/5102/commits/f93fb9e6953e12cd8f5eda6be9cd3dcd61de261f

akyrtzi replied:
On second thought maybe having the change in Swift is appropriate since it's the Swift build that is affected.
But we should at least add a comment to explain that this is due to C++ interop and include the radar link.

eeckstein commented:
@swift-ci smoke test

eeckstein committed:
IRGen: don't enable opaque pointers by default

eeckstein committed:
[rebranch] some test fixes

bnbarham commented on test/SourceKit/CompileNotifications/diagnostics.swift:
```diff
@@ -94,4 +94,6 @@
 // SYNTACTIC: key.syntaxmap:
-// SYNTACTIC_NOT: key.notification: source.notification.compile-did-finish
+
+// TODO: Is this correct?
```
It's not, no. This should be syntactic only and thus shouldn't compile. Feel free to put a radar up and we can take a look at it. I wonder if it's flakey on `main` as well, it would surprise me if there's any rebranch changes that would cause this.

eeckstein replied:
On main this check is ignored because of of the underscore in `SYNTACTIC_NOT` (instead of `SYNTACTIC-NOT`)

bnbarham replied:
Ah wow, totally missed the `_` there. Thanks... that's definitely something for us to look into.
